compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -545,67 +545,109 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
         };
         OperandRef { val, layout }
     }
+
+    pub(crate) fn supports_builder(layout: TyAndLayout<'tcx>) -> bool {
+        match layout.backend_repr {
+            BackendRepr::Memory { .. } if layout.is_zst() => true,
+            BackendRepr::Scalar(_) | BackendRepr::ScalarPair(_, _) => true,
+            _ => false,
+        }
+    }
 }
 
 impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, Result<V, abi::Scalar>> {
     pub(crate) fn insert_field<Bx: BuilderMethods<'a, 'tcx, Value = V>>(
         &mut self,
         bx: &mut Bx,
+        v: VariantIdx,
         f: FieldIdx,
         operand: OperandRef<'tcx, V>,
     ) {
-        let field_layout = self.layout.field(bx.cx(), f.as_usize());
-        let field_offset = self.layout.fields.offset(f.as_usize());
+        let (expect_zst, is_zero_offset) = if let abi::FieldsShape::Primitive = self.layout.fields {
+            // Don't ask for field layout for primitives, because that will panic.
+            if !self.layout.uninhabited {
+                // Real primitives only have one variant, but weird types like
+                // `Result<!, !>` turn out to also be "Primitive", and dead code
+                // like `Err(never)` needs to not ICE.
+                assert_eq!(v, FIRST_VARIANT);
```

Comment:
cogitating...

I wonder about
```rust
enum Never<T> {
    Say(!),
    Again<T>
}
```

The matter of what the "first" variant is probably gets funny when some are uninhabited...